### ISSUE TITLE
Slow SQS task processing due to `wait_time_in_seconds` regression v5.3.2

### ISSUE THREAD

**Issue by dtorres-sf** [celery/kombu#1783]
We recently update from Kombu 5.3.1 to 5.3.2 and saw a very dramatically decrease in the number of tasks that can be processed. Here is an example for a single worker processing a full queue - both graphs in number of messages deleted (processed) a minute.

V5.3.1 (about 500 messages a minute):

![image](https://github.com/celery/kombu/assets/30841716/45378103-ade4-458f-beaa-53a0e7510a47)

Here is the same exact worker and messages being processed on v5.3.2 (about 5 messages a minute):

![image](https://github.com/celery/kombu/assets/30841716/4ff80ce6-b44b-4d8b-94c4-660b1659631c)

We traced this down to just the upgrade of Kombu. When diving into detail we found that the throughput limitation on 5.3.2 was solely due to one setting `wait_time_in_seconds`. When we looked into the logs we found that after a message was processed (which only took 100ms) the worker sat idle for 9.9 seconds, before processing the next task, even if the queue is completely full. Sure enough the `wait_time_in_seconds` was set to 10 seconds. On 5.3.1 Kombu correctly realizes there is more on the queue and doesn't wait the `wait_time_in_seconds` to process the next message.

We are using Amazon SQS, celery[sqs] v5.2.7, here are the broker settings:
```
CELERY_BROKER_TRANSPORT_OPTIONS = {
    "region": "us-west-2",
    "visibility_timeout": 14400,
    "polling_interval": 1,
    'wait_time_seconds': 10
}
```

Let me know if there is anything else I can provide.

**Comment by vifelisberto** [comment]
We're experiencing the same issue, our worker was extremely slow after updating and we weren't figuring out what it was.

Until I checked this past issue https://github.com/celery/celery/issues/4995 where I commented on the `kombu` version, I believe the problem has happened again

**Comment by auvipy** [comment]
were you able to track down the commit causing this? can you also try celery v5.3.4 with the latest kombu release, please? and let us know

**Comment by dtorres-sf** [comment]
@auvipy I have tried using celery 5.3.4. The celery upgrade did not fix the issue. I confirmed the issue still exists on 5.3.4.

I have not been able to narrow down the exact change that caused the issue but the suspect PR is this one:

https://github.com/celery/kombu/pull/1759/files

This is purely based off of code review, since a lot of the SQS code path changed as part of this PR.

**Comment by samueltrevena** [comment]
We're experiencing the same problem after bumping to celery 5.3.4 and kombu 5.3.2. We first observed that `sqs.getqueueurl` was replaced by `sqs.receivemessage`. We immediately hit our client `read_timeout` and had to increase it alot. Over the last few days we've consumed thousands of messages and the average latency is about 20 sec, compared to 4ms before upgrading to celery 5.3.4 with kombu 5.3.2. We're now forced to revert back to celery 5.3.1 with kombu 5.3.1.

**Comment by nickdjones** [comment]
Can confirm we received the same problems as @samueltrevena above

**Comment by auvipy** [comment]
we might revert that to temporarily fix it https://github.com/celery/kombu/pull/1799

**Comment by tuky** [comment]
we're also facing a ton a timeout issues with celery 5.3.4 + kombu 5.3.2 + SQS. in addition there also is a `KeyError` on `WorkerShutdown` that we now see everytime we deploy (and therefore terminate) our workers:

![image](https://github.com/celery/kombu/assets/2981274/56409883-627c-4ade-8acc-d661ac55184c)

**Comment by rafidka** [comment]
Those who reported seeing issues after upgrading, is any of you able to share a reproducing code so I can debug it and see what is going on?

**Comment by rafidka** [comment]
I am unable to reproduce the issue so far. I used this rather aggressive code, which schedules 10k tasks each 10 seconds:

## worker.py
```
import random
from datetime import datetime

from celery import Celery

app = Celery('worker', broker='sqs://')

app.conf.update(
    broker_transport_options={
        'region': 'us-west-2',
        'visibility_timeout': 300,
        'polling_interval': 1,
        'predefined_queues': {
            'celery': {
                'url': 'https://sqs.us-west-2.amazonaws.com/<account ID>/celery-test',
            }
        }
    },
    task_acks_late=False,
    worker_concurrency=1,
)

@app.task(acks_late=False)
def add(x, y):
    return x + y
```

## scheduler.py
```
import time

from worker import add, sort_large_array

while True:
    print("Scheduling task...")
    for i in range(10000):
        add.delay(4, 4)
    time.sleep(10)
```

Then, I ran celery with auto scaling of (32, 0)

```
PYTHONPATH=<workspace>/kombu:<workspace>/botocore celery -A worker worker --loglevel INFO --autoscale 32,0
```

I tried kombu with my PR and without it. Below are the graphs:

![image](https://github.com/celery/kombu/assets/442447/dead5afd-2127-4c59-8b9c-b7dcdbad8ca7)

The gap is in the middle is when I stopped the code and reverted my change. So, I don't see a difference before vs after.

I will continue trying to reproduce the issue, but if someone has code to share to reproduce it that would be great.

**Comment by auvipy** [comment]
this is a public project which reports to be affected https://github.com/pypi/warehouse/pull/14475

**Comment by dtorres-sf** [comment]
@rafidka can you add this setting to your broker transport options `'wait_time_in_seconds': 10`. I think this setting is required to reproduce the issue.

**Comment by rafidka** [comment]
> @rafidka can you add this setting to your broker transport options `'wait_time_in_seconds': 10`. I think this setting is required to reproduce the issue.

Sure, I will give it a try.

**Comment by rafidka** [comment]
I have an update on this issue that explains the issue @dtorres-sf reported.

TL;DR - The use of boto3 in #1759 resulted in relying on blocking (synchronous) HTTP requests, which caused the performance issue reported in this issue. Continue reading for detailed discussion and proposals.

`kombu` previously used to craft AWS requests manually as explained in detail in #1726, which resulted in an outage when botocore temporarily changed the default protocol to JSON (before rolling back due to the impact on celery and airflow.) To fix the issue, I submitted #1759, which changes kombu to use boto3 instead of manually crafting AWS requests. This way when boto3 changes the default protocol, kombu won't be impacted.

While working on #1759, I did extensive debugging to understand the multi-threading nature of kombu. What I discovered is that there isn't an actual multi-threading in the true sense of the word, but an event loop that runs on the same thread and process and orchestrate the communication with SQS. As such, it didn't appear to me that there is anything to worry about my change, and the testing I did didn't discover any issue. However, it turns out that while kombu's event loop doesn't have actual multi-threading, its [reliance on pycurl](https://github.com/celery/kombu/blob/main/kombu/asynchronous/http/curl.py#L48) (and thus libcurl) meant that the requests to AWS were being done asynchronously. On the other hand, boto3 requests are always done synchronously, i.e. they are blocking requests.

The above meant that my change introduced blocking on the event loop of kombu. This is fine in most of the cases, since the requests to SQS are pretty fast. However, in the case of using long-polling, a call to SQS's ReceiveMessage can last up to 20 seconds (depending on the user configuration). This, again, isn't usually a problem if there aren't many messages (since delays won't be noticed) or there is a continuous stream of messages (since ReceiveMessage will be able to gather the 10 maximum number of messages quickly without having to wait for the full 20 seconds of WaitTimeSeconds). However, in the case where you have enough number of messages that delays will be noticeable, yet not enough to full the entire maximum number of messages, the delays will be noticeable as is reported in this issue.

To solve this problem, I worked on a non-blocking boto3 client and currently testing it and fixing some issues. Assuming all goes well and it works fine, I will push the changes to my local repository and you can test it.

**Comment by rafidka** [comment]
> We're experiencing the same problem after bumping to celery 5.3.4 and kombu 5.3.2. We first observed that `sqs.getqueueurl` was replaced by `sqs.receivemessage`. We immediately hit our client `read_timeout` and had to increase it alot. Over the last few days we've consumed thousands of messages and the average latency is about 20 sec, compared to 4ms before upgrading to celery 5.3.4 with kombu 5.3.2. We're now forced to revert back to celery 5.3.1 with kombu 5.3.1.

My comment above explains this. The 20 seconds latency you are experiencing is actually the WaitTimeSeconds for long polling.

**Comment by rafidka** [comment]
I tried to solution to attempt at fixing the problem and I would like to get feedback.

## Solution 1

In this solution, I implemented a thread-based async boto communication mechanism such that polling doesn't block the event loop.

https://github.com/rafidka/kombu/commit/1ed996717a62cacdae61f272797bf03c460d6459

### Pros 

* Use `boto3` library to communicate with SQS, so it should be future proof.

### Cons

* Depart from using `pycurl` (which is what we have already done when we merged https://github.com/celery/kombu/pull/1759). The authors of pycurl claims that it is several times faster than Requests library, which I think is the one that botocore uses. I haven't seen any performance issue myself, but I cannot be certain.

* Introduce threading to kombu, and threading is known to have its challenges.

## Solution 2

In this solution, I attempted to take a different approach to solve the issue reported in https://github.com/celery/kombu/issues/1726. Instead of moving to botocore, I simply implemented a kombu-specific `get_response` method that hard-codes the parser to use the XML-based query format.

https://github.com/rafidka/kombu/commit/8eecce6f424caed2e6aca521ac4e721bef20f117

### Pros

* Continue to use `pycurl` and thus we don't have to worry about performance degradation.
* Doesn't introduce threading

### Cons

* Doesn't fundamentally solve the issue original reported in #1726 , and in theory (although I have to verify) future updates to botocore might cause some issues if somehow we end up making a request that expects JSON, yet we are using the XML-based query format..

**Comment by ippeiukai** [comment]
My company's systems had to revert Kombu to be pinned at 5.3.1 for this issue. I'm not comfortable with receiving changes as significant as [PR#1759](https://github.com/celery/kombu/pull/1759) and Solution 1 in a patch release.

How about Solution 2 for a quick patch release and then Solution 1 for the next minor (or major) version?

**Comment by rafidka** [comment]
> How about Solution 2 for a quick patch release and then Solution 1 for the next minor (or major) version?

I think this is a reasonable suggestion, and I personally also prefer it, at least for the short term, but I am curious to hear from others as well.

**Comment by tuky** [comment]
> > How about Solution 2 for a quick patch release and then Solution 1 for the next minor (or major) version?
> 
> I think this is a reasonable suggestion, and I personally also prefer it, at least for the short term, but I am curious to hear from others as well.

I also relate with this. And possibly instead of aiming for solution 2 in the long run it could be preferable to shoot for existing `async` boto libraries.

**Comment by rafidka** [comment]
@auvipy , unless you are against the idea, I would like to go ahead with solution 2 above (a botocore-like `get_response` implementation) for now, and later probably with a more fundamental solution that relies completely on botocore. Let me know your thoughts.

**Comment by auvipy** [comment]
what is the performance gain in 1st approach? seeing the code it seems 2nd approach is better. but then again, do we have to reintroduce anything old or obsolete?

**Comment by rafidka** [comment]
> what is the performance gain in 1st approach? seeing the code it seems 2nd approach is better. but then again, do we have to reintroduce anything old or obsolete?

In my testing, I didn't notice a difference between the 1st approach and the original code (without my previous change.) Having said that, if any, the original code will be faster. It relies on pycurl, which is just a binding to the libcurl C library, so it is practically impossible to have a threading-based Python code to out perform it.

As for the 2nd approach, it is better in that it safer and doesn't change anything low level. But in my opinion, it shouldn't be the final solution. So, we should employ it to address the issue now, then later on a slower pace we could implement a more robust solution, ensuring that we do enough performance testing.

**Comment by auvipy** [comment]
yeah, so lets continue with the 2nd way

**Comment by dtorres-sf** [comment]
Thank you @auvipy and @rafidka and everyone else who helped investigate and fix the issue!

**Comment by auvipy** [comment]
Goin to release a new patch version today
